enzyme/Enzyme/MLIR/Passes/EnzymeBatchDiffPass.cpp

```diff
@@ -0,0 +1,451 @@
+//===- EnzymeBatchDiffPass.cpp - Merge autodiff calls into their batched
+// versions
+//------------ //
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "Passes/EnzymeBatchDiffPass.h"
+#include "Dialect/Ops.h"
+#include "Interfaces/GradientUtilsReverse.h"
+#include "PassDetails.h"
+#include "Passes/Passes.h"
+
+#include "mlir/Dialect/Func/IR/FuncOps.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/Interfaces/FunctionInterfaces.h"
+
+#define DEBUG_TYPE "enzyme-diff-batch"
+#define ENZYME_DBGS llvm::dbgs() << "[" << DEBUG_TYPE << "]"
+
+using namespace mlir;
+using namespace mlir::enzyme;
+using namespace enzyme;
+
+namespace mlir {
+namespace enzyme {
+#define GEN_PASS_DEF_BATCHDIFFPASS
+#include "Passes/Passes.h.inc"
+} // namespace enzyme
+} // namespace mlir
+
+namespace mlir {
+namespace enzyme {
+namespace batchutils {
+
```

Comment:
Do we really need to copy this?

Reply:
not really but I can fix it up in https://github.com/EnzymeAD/Enzyme/pull/2460 (this branch is the parent to that)